README.md
[![version](https://img.shields.io/badge/version-2.0.0-blue.svg)](./CHANGELOG.md)
[![license](https://img.shields.io/badge/license-MIT-blue.svg)](./LICENSE)
[![Build Status](https://travis-ci.com/elliptical/tcm.svg)](https://travis-ci.com/elliptical/tcm)
[![Coverage Status](https://coveralls.io/repos/github/elliptical/tcm/badge.svg?branch=develop)](https://coveralls.io/github/elliptical/tcm?branch=develop)

# tcm (Test Case Meta)
This is primarily an excercise in Python metaprogramming which also lets me see GitHub, CI tools, and PyPI in action.

Things to develop:

- a class method decorator to hold a table of arguments
- a metaclass to automatically generate multiple test methods out of each decorated sample method

Tools to use:

- `tox` to test the code with different Python versions
- `pylint` and `flake8` to keep individual commits clean
- `coverall` to ensure 100% code coverage


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.0] - 2021-02-15
### Added
- `Python 3.8` is now supported.
- `Python 3.9` is now supported.

### Removed
- `Python 3.5` is no longer supported.

### Fixed
- `python path/setup.py install` no longer throws an exception trying to
  locate the package's `__init__.py` file.

## [1.0.2] - 2019-02-24
### Changed
- The packaged version of `README.md` no longer includes the badges row.

## [1.0.1] - 2019-02-24
### Changed
- `README.md` is now used instead of `README.rst` for PyPI uploads.

## [1.0.0] - 2019-02-23
### Added
- `Python 3.7` is now supported.

### Removed
- `Python 3.4` is no longer supported.

## [0.1.0] - 2017-05-07
- Initial release.

[Unreleased]: https://github.com/elliptical/tcm/compare/2.0.0...HEAD
[2.0.0]: https://github.com/elliptical/tcm/compare/1.0.2...2.0.0
[1.0.2]: https://github.com/elliptical/tcm/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/elliptical/tcm/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/elliptical/tcm/compare/0.1.0...1.0.0
[0.1.0]: https://github.com/elliptical/tcm/releases/tag/0.1.0


test_tcm.py
import unittest

import tcm


class TestCase(unittest.TestCase):
    def test_module_docstring_is_present(self):
        self.assertIsNotNone(tcm.__doc__)


test/test_decorator.py
import unittest

import tcm
from tcm.decorator import ATTR_NAME


class DecoratorTestCase(unittest.TestCase):
    def test_missing_parentheses_will_normally_raise(self):
        with self.assertRaises(tcm.DecoratorException) as cm:
            @tcm.values
            def test():  # pylint: disable=unused-variable
                pass  # pragma: no cover

        self.assertEqual(
            cm.exception.args[0],
            'Invalid use without parentheses or with a callable as the only argument')

    def test_missing_parentheses_will_not_raise_for_determined_users(self):
        @tcm.values
        @_zilch
        def test():
            pass  # pragma: no cover

        self.assertIsInstance(test, tcm.values)

    def test_missing_arguments_are_stored(self):
        @tcm.values()
        def test():
            pass  # pragma: no cover

        self.assertNotIsInstance(test, tcm.values)
        args, kwargs = getattr(test, ATTR_NAME)
        self.assertTupleEqual(args, ())
        self.assertDictEqual(kwargs, {})

    def test_single_positional_noncallable_argument_is_stored(self):
        @tcm.values('abc')
        def test():
            pass  # pragma: no cover

        self.assertNotIsInstance(test, tcm.values)
        args, kwargs = getattr(test, ATTR_NAME)
        self.assertTupleEqual(args, ('abc',))
        self.assertDictEqual(kwargs, {})

    def test_single_positional_callable_argument_will_raise(self):
        with self.assertRaises(tcm.DecoratorException) as cm:
            @tcm.values(_dummy_callable)
            def test():  # pylint: disable=unused-variable
                pass  # pragma: no cover

        self.assertEqual(
            cm.exception.args[0],
            'Invalid use without parentheses or with a callable as the only argument')

    def test_callable_argument_with_another_positional_argument_are_stored(self):
        @tcm.values(_dummy_callable, 12)
        def test():
            pass  # pragma: no cover

        self.assertNotIsInstance(test, tcm.values)
        args, kwargs = getattr(test, ATTR_NAME)
        self.assertTupleEqual(args, (_dummy_callable, 12))
        self.assertDictEqual(kwargs, {})

    def test_callable_argument_with_a_keyword_argument_are_stored(self):
        @tcm.values(_dummy_callable, kw=None)
        def test():
            pass  # pragma: no cover

        self.assertNotIsInstance(test, tcm.values)
        args, kwargs = getattr(test, ATTR_NAME)
        self.assertTupleEqual(args, (_dummy_callable,))
        self.assertDictEqual(kwargs, {'kw': None})

    def test_single_keyword_noncallable_argument_is_stored(self):
        @tcm.values(kw=None)
        def test():
            pass  # pragma: no cover

        self.assertNotIsInstance(test, tcm.values)
        args, kwargs = getattr(test, ATTR_NAME)
        self.assertTupleEqual(args, ())
        self.assertDictEqual(kwargs, {'kw': None})

    def test_single_keyword_callable_argument_is_stored(self):
        @tcm.values(kw=_dummy_callable)
        def test():
            pass  # pragma: no cover

        self.assertNotIsInstance(test, tcm.values)
        args, kwargs = getattr(test, ATTR_NAME)
        self.assertTupleEqual(args, ())
        self.assertDictEqual(kwargs, {'kw': _dummy_callable})

    def test_mixed_arguments_are_stored(self):
        @tcm.values('abc', 12, kw=None)
        def test():
            pass  # pragma: no cover

        self.assertNotIsInstance(test, tcm.values)
        args, kwargs = getattr(test, ATTR_NAME)
        self.assertTupleEqual(args, ('abc', 12))
        self.assertDictEqual(kwargs, {'kw': None})

    def test_noncallable_object_will_raise(self):
        with self.assertRaises(tcm.DecoratorException) as cm:
            @tcm.values()
            @_zilch
            def test():  # pylint: disable=unused-variable
                pass  # pragma: no cover

        self.assertEqual(cm.exception.args[0], 'The object must be callable')

    def test_unexpected_name_prefix_will_raise(self):
        with self.assertRaises(tcm.DecoratorException) as cm:
            @tcm.values()
            def func():  # pylint: disable=unused-variable
                pass  # pragma: no cover

        self.assertEqual(cm.exception.args[0], 'The object name must start with "test"')

    def test_multiple_decorators_will_raise(self):
        with self.assertRaises(tcm.DecoratorException) as cm:
            @tcm.values()
            @tcm.values()
            def test():  # pylint: disable=unused-variable
                pass  # pragma: no cover

        self.assertEqual(cm.exception.args[0], 'Cannot decorate the same object more than once')

    def test_already_existing_attribute_will_raise(self):
        def test():
            pass  # pragma: no cover

        setattr(test, ATTR_NAME, None)

        with self.assertRaises(tcm.DecoratorException) as cm:
            test = tcm.values()(test)

        self.assertEqual(cm.exception.args[0], 'The object already has the "tcm values" attribute')


def _zilch(_func):
    return None


def _dummy_callable():
    pass    # pragma: no cover


test/test_metaclass.py
import inspect
import unittest

import tcm
from tcm.decorator import ATTR_NAME


class MetaclassTestCase(unittest.TestCase):
    # pylint: disable=no-member

    def test_nondecorated_test_methods_keep_their_names(self):
        class GeneratedTestCase(tcm.TestCase):
            def test_no_attr(self):
                pass  # pragma: no cover

            def test_user_attr(self):
                pass  # pragma: no cover

            setattr(test_user_attr, ATTR_NAME, None)

        self.assertFalse(hasattr(GeneratedTestCase.test_no_attr, ATTR_NAME))
        self.assertTrue(hasattr(GeneratedTestCase.test_user_attr, ATTR_NAME))

    def test_decorated_test_methods_are_replaced_with_generated_ones(self):
        class GeneratedTestCase(tcm.TestCase):
            @tcm.values(*range(9), kw=-1)
            def test_few(self, value):
                """Dummy docstring."""
                self.assertIsInstance(value, int)
                return value

            @tcm.values(*range(10))
            def test_many(self, value):
                self.assertIsInstance(value, int)
                return value

        self.assertFalse(hasattr(GeneratedTestCase, 'test_few'))
        self.assertFalse(hasattr(GeneratedTestCase, 'test_many'))

        self.assertFalse(hasattr(GeneratedTestCase.test_few_1, ATTR_NAME))
        self.assertFalse(hasattr(GeneratedTestCase.test_few_9, ATTR_NAME))
        self.assertFalse(hasattr(GeneratedTestCase.test_few_kw, ATTR_NAME))
        self.assertFalse(hasattr(GeneratedTestCase.test_many_01, ATTR_NAME))
        self.assertFalse(hasattr(GeneratedTestCase.test_many_10, ATTR_NAME))

        self.assertEqual(GeneratedTestCase.test_few_2.__doc__, 'Dummy docstring.')
        self.assertIsNone(GeneratedTestCase.test_many_01.__doc__)

        self.assertEqual(GeneratedTestCase.test_few_1.__name__, 'test_few_1')
        self.assertEqual(GeneratedTestCase.test_few_9.__name__, 'test_few_9')
        self.assertEqual(GeneratedTestCase.test_few_kw.__name__, 'test_few_kw')
        self.assertEqual(GeneratedTestCase.test_many_01.__name__, 'test_many_01')
        self.assertEqual(GeneratedTestCase.test_many_10.__name__, 'test_many_10')

        gtc = GeneratedTestCase()

        self.assertEqual(gtc.test_few_1(), 0)
        self.assertEqual(gtc.test_few_9(), 8)
        self.assertEqual(gtc.test_few_kw(), -1)
        self.assertEqual(gtc.test_many_01(), 0)
        self.assertEqual(gtc.test_many_10(), 9)

    def test_values_are_passed_as_is_for_single_arg_test_method(self):
        class GeneratedTestCase(tcm.TestCase):
            @tcm.values(
                tuple(),
                list(),
                dict(),
            )
            def test(self, x):
                self.assertIsInstance(x, (tuple, list, dict))
                return x

        gtc = GeneratedTestCase()

        self.assertTupleEqual(gtc.test_1(), ())
        self.assertListEqual(gtc.test_2(), [])
        self.assertDictEqual(gtc.test_3(), {})

    def test_values_are_unpacked_for_pos_args_test_method(self):
        class GeneratedTestCase(tcm.TestCase):
            @tcm.values(
                ('a', 1),
                ['bb', 2],
                {'x': 'ccc', 'y': 3},
            )
            def test(self, x, y):
                self.assertIsInstance(x, str)
                self.assertIsInstance(y, int)
                return x, y

        gtc = GeneratedTestCase()

        self.assertTupleEqual(gtc.test_1(), ('a', 1))
        self.assertTupleEqual(gtc.test_2(), ('bb', 2))
        self.assertTupleEqual(gtc.test_3(), ('ccc', 3))

    def test_values_are_unpacked_for_star_args_test_method(self):
        class GeneratedTestCase(tcm.TestCase):
            @tcm.values(
                ('a',),
                ['bb', 2],
                ('ccc', 3, None),
            )
            def test(self, *args):
                self.assertIsInstance(args[0], str)
                return args

        gtc = GeneratedTestCase()

        self.assertTupleEqual(gtc.test_1(), ('a',))
        self.assertTupleEqual(gtc.test_2(), ('bb', 2))
        self.assertTupleEqual(gtc.test_3(), ('ccc', 3, None))

    def test_values_are_unpacked_for_keyword_args_test_method(self):
        class GeneratedTestCase(tcm.TestCase):
            @tcm.values(
                {'x': 'a', 'y': 1, 'z': 0},
                {'x': 'bb', 'y': 2},
            )
            def test(self, z='?', **kwargs):
                x = kwargs['x']
                y = kwargs['y']
                self.assertIsInstance(x, str)
                self.assertIsInstance(y, int)
                return x, y, z

        gtc = GeneratedTestCase()

        self.assertTupleEqual(gtc.test_1(), ('a', 1, 0))
        self.assertTupleEqual(gtc.test_2(), ('bb', 2, '?'))

    def test_incompatible_test_method_signature_will_raise_upon_test_run(self):
        class GeneratedTestCase(tcm.TestCase):
            # Each tcm.values item must be a list, a tuple, or a dict because
            # the test method takes two arguments (not a single test argument).
            @tcm.values(
                ['abc'],
            )
            def test_missing(self, x, y):
                pass  # pragma: no cover

            # Each tcm.values item must be a list, a tuple, or a dict because
            # the test method takes no arguments (not a single test argument).
            @tcm.values(
                ['abc'],
            )
            def test_extra(self):
                pass  # pragma: no cover

        gtc = GeneratedTestCase()

        with self.assertRaises(TypeError) as cm:
            gtc.test_missing_1()
        self.assertEqual(
            cm.exception.args[0],
            "test_missing() missing 1 required positional argument: 'y'")

        with self.assertRaises(TypeError) as cm:
            gtc.test_extra_1()
        self.assertEqual(
            cm.exception.args[0],
            'test_extra() takes 1 positional argument but 2 were given')

    def test_incompatible_test_arg_type_will_raise(self):
        with self.assertRaises(tcm.MetaclassException) as cm:
            class _SpoiledTestCase(tcm.TestCase):
                @tcm.values(
                    ('a', 1),
                    2,
                )
                def test(self, x, y):
                    pass    # pragma: no cover

        self.assertEqual(cm.exception.args[0], 'Invalid test arg: 2')

        # Make sure _SpoiledTestCase does not exist.
        self.assertSetEqual(set(locals()), {'self', 'cm'})

    def test_duplicate_test_method_name_will_raise(self):
        with self.assertRaises(tcm.MetaclassException) as cm:
            class _SpoiledTestCase(tcm.TestCase):
                @tcm.values('abc')
                def test(self, value):
                    pass  # pragma: no cover

                def test_1(self):
                    pass  # pragma: no cover

        _source, base = inspect.getsourcelines(
            MetaclassTestCase.test_duplicate_test_method_name_will_raise)
        self.assertEqual(
            cm.exception.args[0],
            f'Duplicate "test_1" attribute at lines {base + 3} and {base + 7}')

        # Make sure _SpoiledTestCase does not exist.
        self.assertSetEqual(set(locals()), {'self', 'cm', '_source', 'base'})


tcm/decorator.py
"""This module provides a decorator to capture test method inputs."""


from collections import namedtuple


ATTR_NAME = 'tcm values'
TEST_METHOD_PREFIX = 'test'


class DecoratorException(Exception):
    """Exception raised on invalid use of the "values" decorator."""


class values():  # noqa: N801 / pylint: disable=invalid-name,too-few-public-methods
    """Parameterized decorator which stores its arguments in the decorated object."""

    def __init__(self, *args, **kwargs):
        """Capture the decorator arguments."""
        if len(args) == 1 and callable(args[0]) and not kwargs:
            raise DecoratorException(
                'Invalid use without parentheses or with a callable as the only argument')
        self.__captured_arguments = _CapturedArguments(args, kwargs)

    def __call__(self, func):
        """Store the captured arguments in the decorated object."""
        if not callable(func):
            raise DecoratorException('The object must be callable')
        if not func.__name__.startswith(TEST_METHOD_PREFIX):
            raise DecoratorException(f'The object name must start with "{TEST_METHOD_PREFIX}"')

        try:
            attr = getattr(func, ATTR_NAME)
        except AttributeError:
            setattr(func, ATTR_NAME, self.__captured_arguments)
        else:
            if isinstance(attr, _CapturedArguments):
                raise DecoratorException('Cannot decorate the same object more than once')
            raise DecoratorException(f'The object already has the "{ATTR_NAME}" attribute')

        return func


def extract_captured_arguments(func):
    """Raise AttributeError for non-decorated "func", return the captured arguments otherwise."""
    captured_arguments = getattr(func, ATTR_NAME)
    if type(captured_arguments) is not _CapturedArguments:  # pylint: disable=unidiomatic-typecheck
        # The attribute was not set by tcm, so effectively it does not exist.
        raise AttributeError
    delattr(func, ATTR_NAME)
    return captured_arguments


_CapturedArguments = namedtuple('CapturedArguments', 'args, kwargs')


tcm/metaclass.py
"""This module provides a metaclass for generating test methods at runtime."""


from collections import OrderedDict
import functools
import inspect

from .decorator import extract_captured_arguments


class MetaclassException(Exception):
    """Exception raised on errors when generating test methods."""


class TestCaseMeta(type):
    """Metaclass based runtime generator of the test methods."""

    @classmethod
    def __prepare__(cls, name, bases, **kwargs):  # pylint: disable=unused-argument
        """Use ordered mapping for the namespace regardless of the Python version."""
        # Note: absent the __prepare__() method, Python 3.6 would use an ordered
        # mapping while prior versions would stick with a regular dict().
        return OrderedDict()

    def __new__(cls, name, bases, mapping):
        """Create the class after expanding the original mapping."""
        new_mapping = dict()
        for key, value in _expanded_mapping(mapping):
            if key in new_mapping:
                existing = _get_starting_line_number(new_mapping[key])
                current = _get_starting_line_number(value)
                raise MetaclassException(
                    f'Duplicate "{key}" attribute at lines {existing} and {current}')
            new_mapping[key] = value
        return super().__new__(cls, name, bases, new_mapping)


def _expanded_mapping(mapping):
    """Iterate the mapping while generating new test methods from decorated ones."""
    for key, value in mapping.items():
        try:
            captured_arguments = extract_captured_arguments(value)
        except AttributeError:
            # Pass non-decorated items unchanged.
            yield key, value
        else:
            # Use the decorated test method as a sample and generate a distinct
            # test method for every captured argument.
            as_is = _has_single_test_param(value)
            for suffix, arg in _uniformly_named_arguments(captured_arguments):
                generated = _generate_test_method(value, arg, as_is)
                generated.__name__ = key + '_' + suffix
                yield generated.__name__, generated


def _has_single_test_param(func):
    sig = inspect.signature(func)
    params = sig.parameters
    if len(params) != 2:
        return False
    _, test_param = params.values()
    return test_param.kind == test_param.POSITIONAL_OR_KEYWORD


def _uniformly_named_arguments(captured_arguments):
    """Iterate the captured arguments as uniform name/value pairs."""
    args, kwargs = captured_arguments

    # For positional arguments, the name is 1-based index padded with
    # leading zeroes to the length of the last index.
    width = len(str(len(args)))
    for i, arg in enumerate(args, 1):
        name = str(i).zfill(width)
        yield name, arg

    # For keyword arguments, the keyword is taken as a name.
    yield from kwargs.items()


def _generate_test_method(func, arg, as_is):
    """Wrap the original test method by supplying the (possibly unpacked) "arg"."""
    if as_is:
        def _wrapper(self):
            return func(self, arg)
    elif isinstance(arg, (tuple, list)):
        def _wrapper(self):
            return func(self, *arg)
    elif isinstance(arg, dict):
        def _wrapper(self):
            return func(self, **arg)
    else:
        raise MetaclassException(f'Invalid test arg: {arg!r}')

    functools.update_wrapper(_wrapper, func)

    return _wrapper


def _get_starting_line_number(func):
    """Return the starting line number of the test method definition."""
    _, start_line_number = inspect.getsourcelines(func)
    return start_line_number


tcm.py
"""This module provides a framework for generating test methods at runtime."""


from collections import namedtuple
from collections import OrderedDict
import functools
import inspect
import unittest


ATTR_NAME = 'tcm values'
TEST_METHOD_PREFIX = 'test'


class DecoratorException(Exception):
    """Exception raised on invalid use of the "values" decorator."""

    pass


class values():  # noqa: N801 / pylint: disable=invalid-name,too-few-public-methods
    """Parameterized decorator which stores its arguments in the decorated object."""

    def __init__(self, *args, **kwargs):
        """Capture the decorator arguments."""
        if len(args) == 1 and callable(args[0]) and not kwargs:
            raise DecoratorException(
                'Invalid use without parentheses or with a callable as the only argument')
        self.__captured_arguments = _CapturedArguments(args, kwargs)

    def __call__(self, func):
        """Store the captured arguments in the decorated object."""
        if not callable(func):
            raise DecoratorException('The object must be callable')
        if not func.__name__.startswith(TEST_METHOD_PREFIX):
            raise DecoratorException(
                'The object name must start with "{}"'.format(TEST_METHOD_PREFIX))

        try:
            attr = getattr(func, ATTR_NAME)
        except AttributeError:
            setattr(func, ATTR_NAME, self.__captured_arguments)
        else:
            if isinstance(attr, _CapturedArguments):
                raise DecoratorException('Cannot decorate the same object more than once')
            else:
                raise DecoratorException(
                    'The object already has the "{}" attribute'.format(ATTR_NAME))

        return func


_CapturedArguments = namedtuple('CapturedArguments', 'args, kwargs')


class MetaclassException(Exception):
    """Exception raised on errors when generating test methods."""

    pass


class TestCaseMeta(type):
    """Metaclass based runtime generator of the test methods."""

    @classmethod
    def __prepare__(mcs, name, bases, **kwargs):  # noqa: N804 pylint: disable=unused-argument
        """Use ordered mapping for the namespace regardless of the Python version."""
        # Note: absent the __prepare__() method, Python 3.6 would use an ordered
        # mapping while prior versions would stick with a regular dict().
        return OrderedDict()

    def __new__(mcs, name, bases, mapping):  # noqa: N804
        """Create the class after expanding the original mapping."""
        new_mapping = dict()
        for key, value in _expanded_mapping(mapping):
            if key in new_mapping:
                existing = _get_starting_line_number(new_mapping[key])
                current = _get_starting_line_number(value)
                raise MetaclassException(
                    'Duplicate "{}" attribute at lines {} and {}'.format(key, existing, current))
            new_mapping[key] = value
        return super().__new__(mcs, name, bases, new_mapping)


def _expanded_mapping(mapping):
    """Iterate the mapping while generating new test methods from decorated ones."""
    for key, value in mapping.items():
        try:
            captured_arguments = _extract_captured_arguments(value)
        except AttributeError:
            # Pass non-decorated items unchanged.
            yield key, value
        else:
            # Use the decorated test method as a sample and generate a distinct
            # test method for every captured argument.
            as_is = _has_single_test_param(value)
            for suffix, arg in _uniformly_named_arguments(captured_arguments):
                generated = _generate_test_method(value, arg, as_is)
                generated.__name__ = key + '_' + suffix
                yield generated.__name__, generated


def _extract_captured_arguments(func):
    """Raise AttributeError for non-decorated "func", return the captured arguments otherwise."""
    captured_arguments = getattr(func, ATTR_NAME)
    if type(captured_arguments) is not _CapturedArguments:  # pylint: disable=unidiomatic-typecheck
        # The attribute was not set by tcm, so effectively it does not exist.
        raise AttributeError
    delattr(func, ATTR_NAME)
    return captured_arguments


def _has_single_test_param(func):
    sig = inspect.signature(func)
    params = sig.parameters
    if len(params) != 2:
        return False
    _, test_param = params.values()
    return test_param.kind == test_param.POSITIONAL_OR_KEYWORD


def _uniformly_named_arguments(captured_arguments):
    """Iterate the captured arguments as uniform name/value pairs."""
    args, kwargs = captured_arguments

    # For positional arguments, the name is 1-based index padded with
    # leading zeroes to the length of the last index.
    width = len(str(len(args)))
    for i, arg in enumerate(args, 1):
        name = str(i).zfill(width)
        yield name, arg

    # For keyword arguments, the keyword is taken as a name.
    yield from kwargs.items()


def _generate_test_method(func, arg, as_is):
    """Wrap the original test method by supplying the (possibly unpacked) "arg"."""
    if as_is:
        def _wrapper(self):
            return func(self, arg)
    elif isinstance(arg, (tuple, list)):
        def _wrapper(self):
            return func(self, *arg)
    elif isinstance(arg, dict):
        def _wrapper(self):
            return func(self, **arg)
    else:
        raise MetaclassException('Invalid test arg: {!r}'.format(arg))

    functools.update_wrapper(_wrapper, func)

    return _wrapper


def _get_starting_line_number(func):
    """Return the starting line number of the test method definition."""
    # Python 3.4 version of getsourcelines() will return the source of the
    # wrapper instead of the wrapped, so we unwrap it first (this is what
    # Python 3.5 and higher do).
    func = inspect.unwrap(func)
    _, start_line_number = inspect.getsourcelines(func)
    return start_line_number


class TestCase(unittest.TestCase, metaclass=TestCaseMeta):
    """Base class to automatically employ the TestCaseMeta metaclass."""

    pass
